x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx

```diff
@@ -0,0 +1,269 @@
+          const geojsonProperties = hoverFeatures[0].properties;
+          // only set state when necessary since event fires so often
+          if (!isEqual(geojsonProperties, hoverState.geojsonProperties)) {
+            const matchedData = data.find(
+              ({ key }) =>
+                geojsonProperties &&
+                key === geojsonProperties[geojsonKeyProperty]
+            );
+            setHoverState({ geojsonProperties, data: matchedData });
+          }
+        } else {
+          setHoverState({});
+        }
+      }
+    },
+    [map, data, hoverState.geojsonProperties, geojsonKeyProperty]
+  );
+
+  const updateHoverStateOnMouseout = useCallback(
+    (event: mapboxgl.MapMouseEvent & mapboxgl.EventData) => {
+      enableScrollZoom.current = false;
+      setHoverState({});
+    },
+    []
+  );
+
+  // When new, memoized handlers are created with useCallback,
+  // these references allow the old handlers to be removed
+  const updateHoverStateOnMousemoveRef = useRef<
+    ((event: mapboxgl.MapMouseEvent & mapboxgl.EventData) => void) | null
+  >(null);
+
+  const updateHoverStateOnMouseoutRef = useRef<
+    ((event: mapboxgl.MapMouseEvent & mapboxgl.EventData) => void) | null
+  >(null);
+
+  // initialization side effect, only runs once
+  useEffect(() => {
+    if (containerRef.current !== null) {
+      // set up Map object
+      const mapboxMap = new Map({
+        attributionControl: false,
+        container: containerRef.current,
+        style:
+          'https://tiles.maps.elastic.co/styles/osm-bright-desaturated/style.json',
+        ...initialMapboxOptionsRef.current
+      });
+      mapboxMap.dragRotate.disable();
+      mapboxMap.touchZoomRotate.disableRotation();
+      mapboxMap.addControl(
+        new NavigationControl({ showCompass: false }),
+        'top-left'
+      );
+
+      // set up Popup object
+      popupRef.current = new Popup({
+        closeButton: false,
+        closeOnClick: false
+      });
+
+      // only scroll zoom when key is pressed
+      const canvasElement = mapboxMap.getCanvas();
+      canvasElement.addEventListener('wheel', controlScrollZoomOnWheel);
+
+      mapboxMap.on('load', () => {
+        mapboxMap.addSource(CHOROPLETH_POLYGONS_SOURCE_ID, {
+          type: 'geojson',
+          data: initialGeojsonSourceRef.current
+        });
+        setMap(mapboxMap);
+      });
+
+      // cleanup function called when component unmounts
+      return () => {
+        canvasElement.removeEventListener('wheel', controlScrollZoomOnWheel);
+      };
+    }
+  }, [controlScrollZoomOnWheel]);
+
+  // side effect swaps the old handlers with new ones when they update
+  useEffect(() => {
+    if (map) {
+      if (
+        updateHoverStateOnMousemoveRef.current &&
+        updateHoverStateOnMouseoutRef.current
+      ) {
+        map.off('mousemove', updateHoverStateOnMousemoveRef.current);
+        map.off('mouseout', updateHoverStateOnMouseoutRef.current);
```

Comment:
Shouldn't this happen during the cleanup phase? Our how do you remove the event handlers when the component unmounts?